<launch>

    <!-- Arguments -->
    <arg name="config_file"   default="$(find livox_ros_driver)/config/livox_lidar_config_multi.json"/>
    <arg name="xfer_format"   default="0"/>
    <arg name="multi_topic"   default="1"/>  <!-- Enable multi‐topic mode -->
    <arg name="msg_frame_id"  default="livox_frame"/>
    <arg name="data_src"      default="0"/>
    <arg name="publish_freq"  default="10.0"/>
    <arg name="output_type"   default="0"/>
    <arg name="rviz_enable"   default="false"/>
    <arg name="rosbag_enable" default="false"/>
    <arg name="lidar_bag"     default="false"/>
    <arg name="imu_bag"       default="false"/>

    <!-- Global parameters: the driver will read these -->
    <param name="xfer_format"      value="$(arg xfer_format)"/>
    <param name="multi_topic"      value="$(arg multi_topic)"/>
    <param name="data_src"         value="$(arg data_src)"/>
    <param name="publish_freq"     type="double" value="$(arg publish_freq)"/>
    <param name="output_data_type" value="$(arg output_type)"/>
    <param name="user_config_path" value="$(arg config_file)"/>
    <param name="enable_lidar_bag" value="$(arg lidar_bag)"/>
    <param name="enable_imu_bag"   value="$(arg imu_bag)"/>

    <!-- Topic remappings for each LiDAR/IMU in multi‐topic mode -->
    <remap from="/livox/lidar_3WEDJA700100021" to="/livox/lidar_farleft_2"/>
    <remap from="/livox/lidar_3WEDJA600100571" to="/livox/lidar_left_57"/>
    <remap from="/livox/lidar_1HDDH3200104431" to="/livox/lidar_mid_65"/>
    <remap from="/livox/lidar_3WEDJA600100751" to="/livox/lidar_right_75"/>

    <remap from="/livox/imu_3WEDJA700100021"   to="/livox/imu_farleft_2"/>
    <remap from="/livox/imu_3WEDJA600100571"   to="/livox/imu_left_57"/>
    <remap from="/livox/imu_1HDDH3200104431"   to="/livox/imu_mid_65"/>
    <remap from="/livox/imu_3WEDJA600100751"   to="/livox/imu_right_75"/>

    <!-- TF transform between base_link and livox_frame -->
    <node pkg="tf" type="static_transform_publisher" name="base_to_livox_lidar"
          args="0 0 0 0 0 0 /base_link /livox_frame 5"/>

    <!-- Start the integrated Livox Driver -->
    <node name="livox_driver"
          pkg="livox_ros_driver"
          type="livox_ros_driver_node"
          required="true"
          output="screen">
        <!-- (Optional) lock to specific NIC if needed -->


        <!-- Broadcast codes for in-process mode switching -->
        <rosparam param="broadcast_codes">
            - 3WEDJA700100021
            - 3WEDJA600100571
            - 1HDDH3200104431
            - 3WEDJA600100751
        </rosparam>
    </node>

    <!-- Optional: Enable RViz -->
    <group if="$(arg rviz_enable)">
        <node name="rviz" pkg="rviz" type="rviz" respawn="true"
              args="-d $(find livox_ros_driver)/config/display_lidar_points_multi.rviz"/>
    </group>

</launch>

